parquet-hadoop/src/main/java/org/apache/parquet/hadoop/MemoryManager.java

```diff
@@ -74,7 +75,7 @@ private void checkRatio(float ratio) {
    * @param writer     the new created writer
    * @param allocation the requested buffer size
    */
-  synchronized void addWriter(InternalParquetRecordWriter<?> writer, Long allocation) {
+  void addWriter(InternalParquetRecordWriter<?> writer, Long allocation) {
     Long oldValue = writerList.get(writer);
     if (oldValue == null) {
```

Comment:
I think here could update concurrently. The `get` and `put` are two methods, it is not an atomic update here.

Reply:
why not `computeIfAbsent`

Reply:
putIfAbsent() and only update the allocation if it was put in the map? It's probably that allocation code which is holding on to the lock for so long. But: that `updateAllocation()` operation is possibly the slow part -and it now has to be thread safe if the callers are no longer locked